### PR TITLE
rustdoc: move some search code into search.js

### PR DESCRIPTION
**Pull request by jsha**
This reduces main.s from 3094 lines to 1587. Also it saves some bytes
of download in the case where search isn't used.

There were a fair number of variables that needed to be accessible in
both main.js and search.js, but I didn't want to put too many symbols in
the global namespace, so I consolidated much of the search-related
state and functions into a new object `window.searchState`.

Demo at https://hoffman-andrews.com/rust/move-search/std/?search=foo

### CODE REVIEW

File: src/librustdoc/html/render/write_shared.rs [hunk 1]
```diff
@@ -408,9 +408,8 @@ pub(super) fn write_shared(
     // with rustdoc running in parallel.
     all_indexes.sort();
     write_crate("search-index.js", &|| {
-        let mut v = String::from("var searchIndex = JSON.parse('{\\\n");
-        v.push_str(&all_indexes.join(",\\\n"));
-        v.push_str("\\\n}');\ninitSearch(searchIndex);");
+        let v = static_files::SEARCH_JS
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/write_shared.rs:411]: Please don't make this change! `search-index.js` is only supposed to contain the search "content", nothing else. Create a new JS file or insert the content into `main.js` but don't change how we handle `search-index.js`.

Comment (jsha) [reply src/librustdoc/html/render/write_shared.rs:411]: Ok! Happy to split this out into `search.js`. Why is `search-index.js` supposed to only contain the search content and nothing else?

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/write_shared.rs:411]: Multiple reason: this is a file different for every crate, meaning that docs.rs cannot store it once and for all. Also it's faster to parse a smaller file (even if here, the part you add might not be that big compared to the search index itself).

So here, it's mostly for docs.rs reason.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
The end result is very nice, thanks a lot! Just one thing to check first: @rust-lang/docs-rs there is new static JS file added. iirc, there is nothing to be done anymore on docs.rs side now but just want to check before approving this PR.

File: src/librustdoc/html/layout.rs [hunk 2]
```diff
@@ -113,7 +113,8 @@ crate fn render<T: Print, S: Print>(
     <section class=\"footer\"></section>\
     {after_content}\
     <div id=\"rustdoc-vars\" data-root-path=\"{root_path}\" data-current-crate=\"{krate}\" \
-       data-search-js=\"{root_path}search-index{suffix}.js\"></div>
+       data-search-index-js=\"{root_path}search-index{suffix}.js\" \
+       data-search-js=\"{root_path}search{suffix}.js\"></div>
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/layout.rs:117]: Oh actually just realized, this should be:

```suggestion
       data-search-js=\"{static_root_path}search{suffix}.js\"></div>
```

**Review by GuillaumeGomez — CHANGES_REQUESTED** [review]

File: src/librustdoc/html/render/write_shared.rs (hunk 1 above)
Comment (Nemo157) [reply src/librustdoc/html/render/write_shared.rs]: And for docs.rs it would be nice to convert it to `.json` at some point :grin:

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/write_shared.rs]: Not possible as is unfortunately, you can load json files for local docs. :3

**Comment by jyn514** [comment]
> here is new static JS file added. iirc, there is nothing to be done anymore on docs.rs side now but just want to check before approving this PR.

Yes, that's fine. You don't need to keep pinging :)

**Comment by GuillaumeGomez** [comment]
> Yes, that's fine. You don't need to keep pinging :)

Promised, it was the last time! :)

**Comment by jsha** [comment]
Thanks for the review! Should I rebase?

**Comment by jyn514** [comment]
@jsha no need, the PR doesn't have conflicts. If people had to rebase on every merge we'd never get any PR merged :laughing:

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks! r=me once the CI is happy. (If you want to rebase, no one will stop you, but it's not required as @jyn514 explained ;) )

**Comment by jsha** [comment]
Looks like CI is done. :-)

**Pull request merged**